luaui/Widgets/Include/select_api.lua

```diff
@@ -0,0 +1,625 @@
+	local index = indexTemp or 1
+	local cmd = spGetCommandQueue(uid, index)
+	if cmd and cmd[index] and cmd[index]["id"] == cmdId then
+		return true
+	end
+	return false
+end
+
+local function isIdle(udef, _udefid, uid)
+	return spGetCommandQueue(uid, 0) == 0
+end
+
+local function stringContains(mainString, searchString)
+	return mainString:find(searchString, 1, true) ~= nil
+end
+
+-- filter
+local function parseFilter(filterDef)
+	local tokens = {}
+	local tokenIndex = 1
+
+	for token in filterDef:gmatch("[^_]+") do
+		table.insert(tokens, token)
+	end
+
+	local function getNextToken()
+		local token = tokens[tokenIndex]
+		tokenIndex = tokenIndex + 1
+		return token
+	end
+
+	local filters = {}
+	local invertIdMatches = nil
+	local idMatchesSet = {}
+
+	while true do
+		local token = getNextToken()
+		local invert = false
+
+		if token == "Not" then
+			invert = true;
+			token = getNextToken()
+		end
+
+		if not token then
+			break
+		end
+
+		-- simple filters
+		if token == "Aircraft" then
+			filters.aircraft = simpleUdefFilter(invert, "canFly")
+		elseif token == "Builder" then
+			filters.builder = invertCurry(invert, function(udef)
+				return isBuilder(udef)
+			end)
+		elseif token == "Buildoptions" then
+			filters.buildOptions = notEmptyUdefFilter(invert, "buildOptions")
+		elseif token == "Building" then
+			filters.building = simpleUdefFilter(invert, "isBuilding")
+		elseif token == "Cloak" then
+			filters.cloak = simpleUdefFilter(invert, "canCloak")
+		elseif token == "Cloaked" then
+			filters.cloaked = invertCurry(invert, function(udef, _, uid)
+				return udef.canCloak and spGetUnitIsCloaked(uid)
+			end)
+		elseif token == "Jammer" then
+			filters.jammer = invertCurry(invert, function(udef)
+				return udef.jammerRadius > 0
+			end)
+		elseif token == "ManualFireUnit" then
+			filters.manualFire = simpleUdefFilter(invert, "canManualFire")
+		elseif token == "Radar" then
+			filters.radar = invertCurry(invert, function(udef)
+				return udef.radarRadius > 0 or udef.sonarRadius > 0
+			end)
+		elseif token == "Resurrect" then
+			filters.resurrect = simpleUdefFilter(invert, "canResurrect")
+		elseif token == "Stealth" then
+			filters.stealth = simpleUdefFilter(invert, "stealth")
+		elseif token == "Transport" then
+			filters.transport = simpleUdefFilter(invert, "isTransport")
+		elseif token == "Weapons" then
+			filters.weapons = notEmptyUdefFilter(invert, "weapons")
+
+			-- command queue filters
+		elseif token == "Idle" then
+			filters.idle = invertCurry(invert, isIdle)
+		elseif token == "Guarding" then
+			filters.guarding = invertCurry(invert, function(_udef, _udefid, uid)
+				return checkCmd(uid, CMD_GUARD)
+			end)
+		elseif token == "Waiting" then
+			filters.waiting = invertCurry(invert, function(_udef, _udefid, uid)
+				return checkCmd(uid, CMD_WAIT)
+			end)
+		elseif token == "Patrolling" then
+			filters.patrolling = invertCurry(invert, function(_udef, _udefid, uid)
+				for i = 1, 4, 1 do
```

Comment:
I don't know the reasoning, I just wanted to match the C++ code. I got the impression patrol might be implemented with a queue of sub commands. I looked into it a little bit at the time, but I wrote this code 3 months ago. I remember when I was debugging there always being a few commands in queue when a unit was patrolling, but I was looking at a few different things at the time, so not really sure.

Reply:
Add a comment saying why, but also see if this is actually needed, nothing worse than cargo cult programming.
```suggestion
				-- First four positions in queue checked because that is what the engine does
				for i = 1, 4, 1 do
```

Reply:
added the comment and did some testing. The patrol command is normally the second command in the queue, not the first command. This is when it is just patrolling, no other inputs. I don't know why it needs to check the first third and fourth positions in the queue, but it's safe to say that the patrol command is implemented using the command queue, which is the reason for this logic.